### ISSUE TITLE
Question on dataset versioning

### ISSUE THREAD

**Issue by MaxGrtz** [webdataset/webdataset#146]
I was thinking about clever data versioning approaches when working with sharded tar archives. 

I would like to have my original dataset stored in the webdataset format in multiple shards. 
When I change the dataset, e.g. offline balancing or augmentation and what not, I would like to efficiently store only the changes to the original dataset, instead of the whole new dataset. This is for example done by weights and biases, but I couldn't find how exactly they do it. Probably not working on tar archives at all and then just using diff and patch. 

Did you guys foresee anything in this direction or have some useful suggestions? 
basically the problem is to apply diff and patch operations on sharded (possibly gziped) tar archives. 

I am not very experience in working with tar archives and just came across this project a few days ago.
I would appreciate any feedback or ideas. 

Thanks in advance!

**Comment by tmbdev** [comment]
Generally, by far the easiest and most reliable way of transforming and versioning datasets is to actually generate an entirely new dataset. That will ensure the best possible I/O speed. Storage is not that expensive, and it greatly reduced the possibility of errors.

But you can diff and patch a WebDataset easily. Let's say you store your updated records in a standard Python shelve database. Then you can write something like:

```Python

patches = shelve.open("patches")

def patch_records(sample):
    replacement = patches.get(sample["__key__"])
    return replacement if replacement is not None else sample

dataset = WebDataset(...).map(patch_records)
...
```

**Comment by MaxGrtz** [comment]
thanks a lot!
